### PR TITLE
Replace math/rand with crypto/rand in cuckoo

### PR DESCRIPTION
**Pull request by jvsg**
Discussion here: https://app.clickup.com/t/1trucm8

### CODE REVIEW

File: internal/cuckoo/cuckoo.go [hunk 1]
```diff
@@ -181,7 +182,9 @@ func (c *Cuckoo) tryAdd(idx uint64, bucketIndices [Nhash]uint64, ignore bool, ex
 func (c *Cuckoo) tryGreedyAdd(idx uint64, bucketIndices [Nhash]uint64) (homeLessItem uint64, added bool) {
 	for i := 1; i < ReInsertLimit; i++ {
 		// select a random slot to be evicted
-		evictedHIdx := rand.Int31n(Nhash)
+		evictedHIdxByte := make([]byte, 8)
+		rand.Read(evictedHIdxByte)
+		evictedHIdx := binary.BigEndian.Uint64(evictedHIdxByte) % Nhash
```
Comment (juanli16) [inline internal/cuckoo/cuckoo.go:187]:
```suggestion
		evictedHIdxByte := make([]byte, 1)
		rand.Read(evictedHIdxByte)
		evictedHIdx := uint64(evictedHIdxByte[0]) % Nhash
```

**Review by juanli16 — COMMENTED** [review]
It is our fault that we removed the initial seeding of the `math/rand` at some point.
Have you tested seeding `math/rand` with pseudorandom bytes read with `crypto/rand`?

File: internal/cuckoo/cuckoo.go [hunk 2]
```diff
@@ -33,6 +35,18 @@ type CuckooHasher struct {
...
+	if _, err := crand.Read(rb[:]); err != nil {
+		panic(err)
+	}
+
+	// WARNING: math/rand is not concurrency-safe
+	// replace with crypto/rand if that's what you want
+
+	// seed math/rand with crypto/rand
+	rand.Seed(int64(binary.LittleEndian.Uint64(rb[:])))
+
```
Comment (juanli16) [inline internal/cuckoo/cuckoo.go:49]: move this to `NewCuckoo`, as the randomness is only needed for those that insert into the cuckoo hash table.

Comment (jvsg) [reply internal/cuckoo/cuckoo.go]: good catch. Fixed it.

**Review by juanli16 — APPROVED** [review]
Good catch!

**Comment by fsaintjacques** [comment]
Some notes:

- rand.IntX() family of functions are thread-safe, if you jump to the implementation, you'll note that it's wrapped in a `lockedSource` which uses a mutex.
- You almost never want to use the global object. You can create you own private `rand.NewSource` & `rand.New(src)`. This is extremely useful when you want to unit test, you can use a static seed where-as in production you derive the seed from crypto/rand.

TLDR, rewrite Cuckoo such that it stashes his own rand.Rand.
